### PR TITLE
remove WCHAR which doesn't work really in OMI

### PR DESCRIPTION
**Pull request by BingyuLiu**
@palladia, make this code change to fix  issue  https://github.com/Microsoft/omi/issues/460

@jeffaco  and @paulcallen , I think you both know this issue also, please also take a look at these changes through you doesn't work for OMI already. Thank you in advance!

I will make a sepertally code change to coordinate these unit test cases that need to be changed.

### CODE REVIEW

**Comment by BingyuLiu** [comment]
ping @palladia

File: Unix/common/windows/config.h [hunk 1]
```diff
@@ -34,7 +34,6 @@
 #endif
 
 #define CONFIG_HAVE_FUNCTION_MACRO 1
-#define CONFIG_ENABLE_WCHAR 1
 #define CONFIG_HAVE_ATOMIC_INTRINSICS 1
 #define CONFIG_HAVE_WCSTOLL 1
```
Comment (palladia) [inline Unix/common/windows/config.h]: You probably want to also get rid of CONFIG_HAVE_WCSTOLL, CONFIG_HAVE_WCSCASECMP, WCSLEN, CONFIG_HAVE_WCSDUP also

**Comment by palladia** [comment]
A few other things to check:

1.  The "configure" shell script has wide-char support

2.  MI_CHAR_TYPE of "2" means wide-char support

**Comment by BingyuLiu** [comment]
Thanks, I will take look at them.

**Comment by BingyuLiu** [comment]
@palladia , I have made more changes, please take a look at them. Thanks!

File: Unix/configure [hunk 2]
```diff
@@ -1904,24 +1904,6 @@ else
     echo "/* #define CONFIG_HAVE_SCHED_GETCPU */" >> $fn
 fi
 
-if [ "$have_wcstoll" = "1" ]; then
-    echo "#define CONFIG_HAVE_WCSTOLL" >> $fn
-else
-    echo "/* #define CONFIG_HAVE_WCSTOLL */" >> $fn
-fi
-
```
Comment (palladia) [inline Unix/configure]: You should remove code that tests for wcstoll too.

File: Unix/doc/wide-character-support.txt [hunk 3]
```diff
@@ -5,7 +5,6 @@ To build OMI to support wide characters, configure like this:
 This enables the use of wide characters in the code and defines the following 
 macros:
 
-    CONFIG_ENABLE_WCHAR
     MI_CHAR_TYPE==2
```
Comment (palladia) [inline Unix/doc/wide-character-support.txt]: You should just remove the whole document.

File: Unix/nits/base/nits.h.bkp [hunk 4]
```diff
@@ -873,23 +873,13 @@ NITS_EXPORT ptrdiff_t NITS_PRESENCE;
```
Comment (palladia) [inline Unix/nits/base/nits.h.bkp]: Why do we need a backup copy of nits.h?  It looks like Bruce created this.  Check with him, and delete this file.

File: Unix/nits/injector/injector.cpp.bkp [hunk 5]
```diff
@@ -126,24 +126,14 @@ PWSTR ConvertStringToW(_In_opt_z_ const char *buf)
```
Comment (palladia) [inline Unix/nits/injector/injector.cpp.bkp]: Again, shouldn't have backup copies.  Delete.

File: Unix/pal/pal [hunk 6]
```diff
@@ -210,11 +210,7 @@
 **==============================================================================
```
Comment (palladia) [inline Unix/pal/pal]: This looks like an old version of palcommon.h.    What happens if you delete this file?

File: Unix/nits/base/nits.h.bkp (hunk 4 above)
Comment (BingyuLiu) [reply Unix/nits/base/nits.h.bkp]: I will talk to Bruce to know why.

**Comment by palladia** [comment]
I think you are missing some files still.  I did a quick count, and see that there are 53 files with "WCHAR" in it.

Please re-check.  I see many tests have WCHAR sections.

**Comment by BingyuLiu** [comment]
1. I will make a separate commit to modify these related unit test cases.
2. like the following code in nits.h, how can I delete the wchar, what data type should I use after delete wchar_t?

    NitsResult (NITS_CALL *CompareStringW)(
        _In_z_ const wchar_t *lhs,
        _In_z_ const wchar_t *rhs,
        _In_z_ const char *lhsText,
        _In_z_ const char *rhsText,
        _In_z_ const wchar_t *description,
               NitsCallSite line,
               NitsFaultMode mode);

File: Unix/doc/wide-character-support.txt (hunk 3 above)
Comment (BingyuLiu) [reply Unix/doc/wide-character-support.txt]: removed it

**Comment by palladia** [comment]
Nits functions have the A version and the W version.  A is for ASCII and W is for wide?  We probably don't need the W version of functions.  

I am not too worried about nits.  If it's too much work, don't worry about it for now.

**Comment by BingyuLiu** [comment]
I will make a separate commit to remove Nits function W version if we don't need it.

File: Unix/tests/codec/mof/blue/test_lex.cpp [hunk 7]
```diff
@@ -28,20 +28,20 @@
 
 #define TEST_ASSERT(x) NitsAssert(x, MI_T(""))
 
-#if (defined(sun) || defined(hpux) || defined(aix)) && defined(CONFIG_ENABLE_WCHAR)
+#if defined(sun) || defined(hpux) || defined(aix)
```
Comment (palladia) [inline Unix/tests/codec/mof/blue/test_lex.cpp]: Why are you doing this?  Why not delete section?

File: Unix/tests/codec/mof/blue/test_lex.cpp [hunk 8]
```diff
@@ -28,20 +28,20 @@
...
 
 // 
 // This failes due to difficulty supporting four byte character sets in codec/mof/parser/utility.c
 // An issue has been filed.
 //
-#define DISABLE_WCHAR_TESTS 1
+#define DISABLE_CHAR_TESTS 1
 #endif
```
Comment (palladia) [inline Unix/tests/codec/mof/blue/test_lex.cpp]: Why define a new variable?

File: Unix/tests/pal/test_pal.cpp [hunk 9]
```diff
@@ -32,14 +32,14 @@
 #include <pal/hashmap.h>
 #include <pal/format.h>
 
-#if (defined(sun) || defined(hpux) || defined(aix)) && defined(CONFIG_ENABLE_WCHAR)
+#if defined(sun) || defined(hpux) || defined(aix)
```
Comment (palladia) [inline Unix/tests/pal/test_pal.cpp]: Again, if CONFIG_ENABLE_WCHAR is never defined, then this if should always fail.  Note it's &&, not ||

File: Unix/tests/codec/mof/blue/test_lex.cpp (hunk 8 above)
Comment (BingyuLiu) [reply Unix/tests/codec/mof/blue/test_lex.cpp]: In line 44 of file test_lex.cpp, the DISABLE_CHAR_TESTS are never defined.  I think there
 is typo in line 37, it should be DISABLE_CHAR_TESTS, but not DISABLE_WCHAR_TESTS.

File: Unix/nits/injector/injector.cpp.bkp (hunk 5 above)
Comment (BingyuLiu) [reply Unix/nits/injector/injector.cpp.bkp]: deleted it.

File: Unix/tests/pal/test_pal.cpp (hunk 9 above)
Comment (BingyuLiu) [reply Unix/tests/pal/test_pal.cpp]: you are right. I have deleted this section.

**Comment by BingyuLiu** [comment]
Hi George,

Thanks for your comments.  I have fixed these issues as your comments. And except AIX7, all platforms run pass in pbuild.

Thanks,
John Liu

File: Unix/pal/format.placeholders.h [hunk 10]
```diff
@@ -9,17 +9,8 @@
...
-#ifdef INTLSTR_RC_INC
 # define PAL_PRItstr "s"
 # define PAL_PRItchr "c"
-#elif defined(CONFIG_ENABLE_WCHAR)
-# define PAL_PRItstr "S"
-# define PAL_PRItchr "C"
-#else
-# define PAL_PRItstr "s"
-# define PAL_PRItchr "c"
-#endif
```
Comment (palladia) [inline Unix/pal/format.placeholders.h]: You shouldn't remove #ifdef INTLSTR_RC_INC.

File: Unix/pal/strings.c [hunk 11]
```diff
@@ -348,10 +306,9 @@ int wcscasecmp(const wchar_t* s1, const wchar_t* s2)
 
     return (int)((unsigned int)*s1 - (unsigned int)*s2);    
 }
-#endif
```
Comment (palladia) [inline Unix/pal/strings.c]: Can you check to see if anyone actually uses wcsncasecmp()?  If not, just delete it.  Without wchar, there shouldn't be anyone calling this function.

File: Unix/pal/strings.c [hunk 12]
```diff
@@ -476,121 +433,4 @@ long long wcstoll(const wchar_t* s, wchar_t** e, int b)
     }
     return (long long)value;
 }
-#endif
-
```
Comment (palladia) [inline Unix/pal/strings.c]: Same comment as wcsncasecmp().  In fact, you shouldn't add wctype.h.

File: Unix/tests/codec/mof/blue/qualifierset.c [hunk 13]
```diff
@@ -16,10 +16,8 @@
 #else
 #include <wchar.h>
 
-#if !defined(CONFIG_HAVE_WCSCASECMP)
 // defined in pal/strings
 int wcscasecmp(const wchar_t* s1, const wchar_t* s2);
-#endif
```
Comment (palladia) [inline Unix/tests/codec/mof/blue/qualifierset.c]: Like above, do we really needs wcscasecmp()?

File: Unix/tests/pal/test_pal.cpp (hunk 9 above)
Comment (palladia) [reply Unix/tests/pal/test_pal.cpp]: You didn't delete?

**Comment by BingyuLiu** [comment]
Ping....

File: Unix/nits/base/Run.cpp [hunk 14]
```diff
@@ -1043,7 +1037,7 @@ bool Run::ParseArgument(_In_ wchar_t *arg)
         }
 
         char buf[MAX_PATH];
-        wcstombs(buf, second, MAX_PATH);
+		strcpy(buf, second);
```
Comment (palladia) [inline Unix/nits/base/Run.cpp]: The single-char equivalent of wcstombs() is strlcpy(), not strcpy().  But in this case, there is no reason to make another copy.  You should be able to just use "second", and not bother creating "buf"?

File: Unix/nits/base/nits_c.cpp [hunk 15]
```diff
@@ -100,11 +100,7 @@ static void NITS_CALL TraceA(
            NitsCallSite line,
            NitsFaultMode mode)
 {
-    //TODO: This should be the base logging API that everything else uses.
-    //  The underlying cross-proc buffer and log should be ANSI as well.
-    PWSTR wideText = ConvertStringToW(text);
-    TraceW(wideText, line, mode);
-    delete [] wideText;
+    subTrace(text, line, mode);
```
Comment (palladia) [inline Unix/nits/base/nits_c.cpp]: Why not just rename subTrace() to TraceA() instead?  So you don't have an extra function.

File: Unix/pal/strings.c [hunk 16]
```diff
@@ -330,28 +288,8 @@ int wcsncasecmp(const wchar_t* s1, const wchar_t* s2, size_t n)
...
-        s2++;
-    }
-
-    return (int)((unsigned int)*s1 - (unsigned int)*s2);    
-}
-#endif
-
-#if !defined(CONFIG_HAVE_WCSTOLL)
 #include <errno.h>
+#include <wctype.h>
```
Comment (palladia) [inline Unix/pal/strings.c]: Why add wctype.h here?

File: Unix/common/windows/config.h [hunk 17]
```diff
@@ -34,11 +34,7 @@
 #endif
 
 #define CONFIG_HAVE_FUNCTION_MACRO 1
-#define CONFIG_ENABLE_WCHAR 1
 #define CONFIG_HAVE_ATOMIC_INTRINSICS 1
-#define CONFIG_HAVE_WCSTOLL 1
-#define CONFIG_HAVE_WCSDUP 1
-#define CONFIG_HAVE_WCSCASECMP 1
 #define CONFIG_HAVE_BACKTRACE 1
```
Comment (palladia) [inline Unix/common/windows/config.h]: Do we really need this "windows" file?

**Comment by palladia** [comment]
Note that Travis build failed.  Please fix that too.

File: Unix/pal/strings.c [hunk 18]
```diff
@@ -329,25 +288,6 @@ int wcsncasecmp(const wchar_t* s1, const wchar_t* s2, size_t n)
...
-        s1Char = PAL_tolower(*s1);
-        s2Char = PAL_tolower(*s2);
-        if(s1Char != s2Char)
-            return (int)(s1Char - s2Char);
-        s1++;
-        s2++;
-    }
-
-    return (int)((unsigned int)*s1 - (unsigned int)*s2);    
-}
```
Comment (palladia) [inline Unix/pal/strings.c]: You should put this function,  wcscasecmp() back.

**Comment by BingyuLiu** [comment]
add wcscasecmp function back
pbuild run pass
squash 6 commits to one.

**Comment by BingyuLiu** [comment]
pbuild results:
-rw-rw-r--.  1 johnliu johnliu 1052499 Oct  8 15:40 done-centos_61_64.log
-rw-rw-r--.  1 johnliu johnliu  975456 Oct  8 15:31 done-centos_7_3_64.log
-rw-rw-r--.  1 johnliu johnliu 1020591 Oct  8 15:32 done-centos_7_64.log
-rw-rw-r--.  1 johnliu johnliu  886934 Oct  8 15:40 done-hp_v3_ia64.log
-rw-rw-r--.  1 johnliu johnliu 1008860 Oct  8 15:34 done-redhat_5_32.log
-rw-rw-r--.  1 johnliu johnliu 1057259 Oct  8 15:33 done-redhat_5_64.log
-rw-rw-r--.  1 johnliu johnliu 1037003 Oct  8 15:33 done-redhat_6_32.log
-rw-rw-r--.  1 johnliu johnliu  984782 Oct  8 15:30 done-redhat_7.1_ppc.log
-rw-rw-r--.  1 johnliu johnliu 1006568 Oct  8 15:31 done-redhat_7_64.log
-rw-rw-r--.  1 johnliu johnliu 1141931 Oct  8 15:33 done-sun_5.10_sparc.log
-rw-rw-r--.  1 johnliu johnliu 1022126 Oct  8 15:35 done-sun_5.10_x86.log
-rw-rw-r--.  1 johnliu johnliu  958748 Oct  8 15:35 done-sun_5.11_sparc.log
-rw-rw-r--.  1 johnliu johnliu  915910 Oct  8 15:33 done-sun_5.11_x86.log
-rw-rw-r--.  1 johnliu johnliu  992601 Oct  8 15:32 done-suse_11_32.log
-rw-rw-r--.  1 johnliu johnliu 1011897 Oct  8 15:32 done-suse_11_64.log
-rw-rw-r--.  1 johnliu johnliu 1007409 Oct  8 15:34 done-ubuntu_14_64.log
-rw-rw-r--.  1 johnliu johnliu  987057 Oct  8 15:36 done-ubuntu_16_64.log
-rw-rw-r--.  1 johnliu johnliu 1005607 Oct  8 15:37 done-ubuntu_18_64.log

**Pull request merged**